chapter_recurrent-neural-networks/rnn-concise.ipynb

```diff
@@ -0,0 +1,808 @@
+    "                } else {\n",
+    "                    NDList pairResult;\n",
+    "                    if (state == null) {\n",
+    "                        // Begin state\n",
+    "                        pairResult =\n",
+    "                                ((AbstractBlock) net)\n",
+    "                                        .forward(\n",
+    "                                                new ParameterStore(manager, false),\n",
+    "                                                new NDList(X),\n",
+    "                                                true);\n",
+    "                    } else {\n",
+    "                        pairResult =\n",
+    "                                ((AbstractBlock) net)\n",
+    "                                        .forward(\n",
+    "                                                new ParameterStore(manager, false),\n",
+    "                                                new NDList(X).addAll(state),\n",
+    "                                                true);\n",
+    "                    }\n",
+    "                    yHat = pairResult.get(0);\n",
+    "                    state = pairResult.subNDList(1);\n",
+    "                }\n",
+    "\n",
+    "                NDArray l = loss.evaluate(new NDList(y), new NDList(yHat)).mean();\n",
+    "                gc.backward(l);\n",
+    "                metric.add(new float[] {l.getFloat() * y.size(), y.size()});\n",
+    "            }\n",
+    "            gradClipping(net, 1, childManager);\n",
+    "            updater.apply(1, childManager); // Since the `mean` function has been invoked\n",
+    "        }\n",
+    "    }\n",
+    "    return new Pair<>(Math.exp(metric.get(0) / metric.get(1)), metric.get(1) / watch.stop());\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "/** Clip the gradient. */\n",
+    "public static void gradClipping(Object net, int theta, NDManager manager) {\n",
+    "    double result = 0;\n",
+    "    NDList params;\n",
+    "    if (net instanceof RNNModelScratch) {\n",
+    "        params = ((RNNModelScratch) net).params;\n",
+    "    } else {\n",
+    "        params = new NDList();\n",
+    "        for (Pair<String, Parameter> pair : ((AbstractBlock) net).getParameters()) {\n",
+    "            params.add(pair.getValue().getArray());\n",
+    "        }\n",
+    "    }\n",
+    "    for (NDArray p : params) {\n",
+    "        NDArray gradient = p.getGradient().stopGradient();\n",
+    "        gradient.attach(manager);\n",
+    "        result += gradient.pow(2).sum().getFloat();\n",
+    "    }\n",
+    "    double norm = Math.sqrt(result);\n",
+    "    if (norm > theta) {\n",
+    "        for (NDArray param : params) {\n",
+    "            NDArray gradient = param.getGradient().stopGradient();\n",
+    "            param.getGradient().set(new NDIndex(\":\"), gradient.mul(theta / norm));\n",
```

Comment:
I think previously you use param.getGradient().muli. pick one and stick with it

Reply:
Done, switched it to `muli` as well as for the rest of the uses.